paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/AsyncRefreshLookupTable.java

```diff
@@ -0,0 +1,149 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.lookup;
+
+import org.apache.paimon.data.InternalRow;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.table.FileStoreTable;
+import org.apache.paimon.utils.ExecutorThreadFactory;
+import org.apache.paimon.utils.ExecutorUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.apache.paimon.flink.FlinkConnectorOptions.LOOKUP_REFRESH_ASYNC;
+import static org.apache.paimon.flink.FlinkConnectorOptions.LOOKUP_REFRESH_ASYNC_PENDING_SNAPSHOT_COUNT;
+
+/** A {@link LookupTable} supports async refresh. */
+public abstract class AsyncRefreshLookupTable implements LookupTable {
+    private static final Logger LOG = LoggerFactory.getLogger(AsyncRefreshLookupTable.class);
+    private final FileStoreTable table;
+
+    private final int maxPendingSnapshotCount;
+
+    @Nullable private final ExecutorService refreshExecutor;
+
+    private final AtomicReference<Exception> cachedException;
+
+    private Future<?> refreshFuture;
+
+    protected final Object lock;
+    protected final boolean refreshAsync;
+
+    public AsyncRefreshLookupTable(FileStoreTable table) {
+        Options options = Options.fromMap(table.options());
+        this.table = table;
+        this.refreshAsync = options.get(LOOKUP_REFRESH_ASYNC);
+        this.refreshExecutor =
+                this.refreshAsync
+                        ? Executors.newSingleThreadExecutor(
+                                new ExecutorThreadFactory(
+                                        String.format(
+                                                "%s-lookup-refresh",
+                                                Thread.currentThread().getName())))
+                        : null;
+        this.lock = this.refreshAsync ? new Object() : null;
+        this.cachedException = new AtomicReference<>();
+        this.maxPendingSnapshotCount = options.get(LOOKUP_REFRESH_ASYNC_PENDING_SNAPSHOT_COUNT);
+    }
+
+    @Override
+    public List<InternalRow> get(InternalRow key) throws IOException {
+        List<InternalRow> values;
+        if (refreshAsync) {
+            synchronized (lock) {
```

Comment:
I'm curious about what this lock to protect, Is this to avoid the race condition between `get` and `refresh` ? Or the parallelism `get` ?  
I think we should  ensure the `QueryExecutor` is `ThreadSafe` Or guarded by lock.

Reply:
@Aitozi this is first introduced in `FullCacheLookupTable` scenario. Explained by @FangYongs, 'Currently there will be primaryKeyRow, RocksDBSetState and RocksDBListState in PrimaryKeyLookupTable, SecondaryIndexLookupTable and NoPrimaryKeyLookupTable, they are not thread safe. I think we can keep the lock in get method and optimize it when we find that it has become a bottleneck.' See more in https://github.com/apache/paimon/pull/3297